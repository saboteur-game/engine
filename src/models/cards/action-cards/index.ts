import ActionCard from "./action-card";
import {
  BoardActionCard,
  MapActionCard,
  RockfallActionCard,
} from "./board-action-cards";
import {
  ToolActionCard,
  BreakActionCard,
  RepairActionCard,
} from "./tool-action-cards";

export {
  ActionCard,
  BoardActionCard,
  MapActionCard,
  RockfallActionCard,
  ToolActionCard,
  BreakActionCard,
  RepairActionCard,
};
