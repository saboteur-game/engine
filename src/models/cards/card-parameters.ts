import { Tools } from "../../constants";
import Position from "../position";
import Player from "../player";

export interface IBoardCardParameters {
  position: Position;
}

export interface IToolActionCardParameters {
  player: Player;
  appliedTo: Tools;
}

type CardParameters = IToolActionCardParameters | IBoardCardParameters;

export default CardParameters;
