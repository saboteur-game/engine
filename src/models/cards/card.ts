import { generateId, Pojo } from "../../utils";
import CardParameters from "./card-parameters";

export enum Status {
  unused = "unused",
  played = "played",
  discarded = "discarded",
}

class Card {
  id: string;
  status: Status;

  constructor() {
    this.id = generateId();
    this.status = Status.unused;
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  play(parameters: CardParameters): Card {
    return this;
  }

  setDiscarded(): void {
    // TODO: Check if the state was unused and throw if it was anything else
    this.status = Status.discarded;
  }

  setPlayed(): void {
    // TODO: Check if the state was unused or played and throw if it was anything else
    this.status = Status.played;
  }

  toJS(): Pojo {
    return {
      id: this.id,
      status: this.status,
    };
  }
}

export default Card;
