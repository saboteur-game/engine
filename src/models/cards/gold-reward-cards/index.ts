import { generateId, Pojo } from "../../../utils";

class GoldRewardCard {
  id: string;
  value: number;

  constructor(value: 1 | 2 | 3) {
    this.id = generateId();
    this.value = value;
  }

  toJS(): Pojo {
    return {
      id: this.id,
      value: this.value,
    };
  }
}

export default GoldRewardCard;
