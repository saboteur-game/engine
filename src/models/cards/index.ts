import { shuffle } from "../../utils";
import { Tools } from "../../constants";
import Card, { Sides } from "./card";
import {
  StartPathCard,
  GoldFinishPathCard,
  RockFinishPathCard,
  PassageCard,
  DeadendCard,
  PathCard,
} from "./path-cards";
import {
  MapActionCard,
  RockfallActionCard,
  BreakActionCard,
  RepairActionCard,
  ActionCard,
} from "./action-cards";

const multiply = (length: number, getValue: () => Card) =>
  new Array(length).fill(undefined).map(() => getValue());
const randomBoolean = () => Math.random() < 0.5;

/*
 * Saboteur cards
 * - 44 path cards
 *  -- 1 starting card
 *  -- 3 finishing cards
 *    --- 1 gold [Connected, 1-2-3-4]
 *    --- 2 rock
 *       ---- 1 [Connected 1-2]
 *       ---- 1 [Connected 1-4]
 *  -- 40 tunnel cards
 *    --- 31 passage cards
 *       ---- 3 passage [Connected 2-4]
 *       ---- 4 passage [Connected 1-3]
 *       ---- 4 passage [Connected 1-4]
 *       ---- 5 passage [Connected 1-2]
 *       ---- 5 passage [Connected 1-2-4]
 *       ---- 5 passage [Connected 1-2-3]
 *       ---- 5 passage [Connected 1-2-3-4]
 *    --- 9 dead-end cards
 *       ---- 1 dead-end [Connected 1]
 *       ---- 1 dead-end [Connected 2]
 *       ---- 1 dead-end [Connected 1-2]
 *       ---- 1 dead-end [Connected 1-3]
 *       ---- 1 dead-end [Connected 1-4]
 *       ---- 1 dead-end [Connected 2-4]
 *       ---- 1 dead-end [Connected 1-2-3]
 *       ---- 1 dead-end [Connected 1-2-4]
 *       ---- 1 dead-end [Connected 1-2-3-4]
 * - 27 action cards
 *  -- 6 map cards
 *  -- 3 rock-fall cards
 *  -- 9 break tool cards
 *    --- 3 pick
 *    --- 3 lamp
 *    --- 3 wagon
 *  -- 9 repair tool cards
 *    --- 2 pick
 *    --- 2 lamp
 *    --- 2 wagon
 *    --- 1 pick and lamp
 *    --- 1 pick and wagon
 *    --- 1 lamp and wagon
 *
 * Notes: Card connections
 *
 *        (1)
 *     +-------+
 *     |       |
 * (4) |       | (2)
 *     |       |
 *     +-------+
 *        (3)
 */

export interface IPlacedCards {
  start: StartPathCard;
  gold: GoldFinishPathCard;
  rock1: RockFinishPathCard;
  rock2: RockFinishPathCard;
}

export const getPlacedCards = (): IPlacedCards => {
  const start = new StartPathCard();
  const gold = new GoldFinishPathCard();
  const rock1 = new RockFinishPathCard([Sides.top, Sides.right]);
  const rock2 = new RockFinishPathCard([Sides.top, Sides.left]);

  [start, gold, rock1, rock2].forEach((card) => card.setPlayed());

  return { start, gold, rock1, rock2 };
};

export const getShuffledDeck = (): Array<ActionCard | PathCard> => {
  const pathCards = [
    ...multiply(
      3,
      () => new PassageCard([Sides.right, Sides.left], randomBoolean())
    ),
    ...multiply(
      4,
      () => new PassageCard([Sides.top, Sides.bottom], randomBoolean())
    ),
    ...multiply(
      4,
      () => new PassageCard([Sides.top, Sides.left], randomBoolean())
    ),
    ...multiply(
      5,
      () => new PassageCard([Sides.top, Sides.right], randomBoolean())
    ),
    ...multiply(
      5,
      () =>
        new PassageCard([Sides.top, Sides.right, Sides.left], randomBoolean())
    ),
    ...multiply(
      5,
      () =>
        new PassageCard([Sides.top, Sides.right, Sides.bottom], randomBoolean())
    ),
    ...multiply(
      5,
      () =>
        new PassageCard(
          [Sides.top, Sides.right, Sides.bottom, Sides.left],
          randomBoolean()
        )
    ),
    new DeadendCard([Sides.top], randomBoolean()),
    new DeadendCard([Sides.right], randomBoolean()),
    new DeadendCard([Sides.top, Sides.right], randomBoolean()),
    new DeadendCard([Sides.top, Sides.bottom], randomBoolean()),
    new DeadendCard([Sides.top, Sides.left], randomBoolean()),
    new DeadendCard([Sides.right, Sides.left], randomBoolean()),
    new DeadendCard([Sides.top, Sides.right, Sides.bottom], randomBoolean()),
    new DeadendCard([Sides.top, Sides.right, Sides.left], randomBoolean()),
    new DeadendCard(
      [Sides.top, Sides.right, Sides.bottom, Sides.left],
      randomBoolean()
    ),
  ] as Array<PathCard>;

  const actionCards = [
    ...multiply(6, () => new MapActionCard()),
    ...multiply(3, () => new RockfallActionCard()),
    ...multiply(3, () => new BreakActionCard(Tools.pick)),
    ...multiply(3, () => new BreakActionCard(Tools.lamp)),
    ...multiply(3, () => new BreakActionCard(Tools.wagon)),
    ...multiply(2, () => new RepairActionCard(Tools.pick)),
    ...multiply(2, () => new RepairActionCard(Tools.lamp)),
    ...multiply(2, () => new RepairActionCard(Tools.wagon)),
    new RepairActionCard([Tools.pick, Tools.lamp]),
    new RepairActionCard([Tools.pick, Tools.wagon]),
    new RepairActionCard([Tools.lamp, Tools.wagon]),
  ] as Array<ActionCard>;

  return shuffle(
    ([] as Array<ActionCard | PathCard>).concat(pathCards, actionCards)
  );
};
