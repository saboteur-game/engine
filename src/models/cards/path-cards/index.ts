import PathCard from "./path-card";
import { TunnelCard, PassageCard, DeadendCard } from "./tunnel-cards";
import {
  StartPathCard,
  FinishPathCard,
  GoldFinishPathCard,
  RockFinishPathCard,
} from "./boundary-cards";

export {
  PathCard,
  TunnelCard,
  PassageCard,
  DeadendCard,
  StartPathCard,
  FinishPathCard,
  GoldFinishPathCard,
  RockFinishPathCard,
};
