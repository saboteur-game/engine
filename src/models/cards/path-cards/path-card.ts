import Card, { Sides } from "../card";
import { Pojo } from "../../../utils";
import CardParameters from "../card-parameters";

class PathCard extends Card {
  connectors: Sides[];
  isUpsideDown: boolean;
  parameters: CardParameters | undefined;

  constructor(connectors: Sides[], isUpsideDown?: boolean) {
    super();
    this.connectors = connectors;
    this.isUpsideDown = isUpsideDown || false;
  }

  rotate(): boolean {
    this.isUpsideDown = !this.isUpsideDown;
    return this.isUpsideDown;
  }

  play(parameters: CardParameters): PathCard {
    super.play(parameters);
    this.parameters = parameters;
    return this;
  }

  visualize(): string {
    throw new Error("Not implemented");
  }

  toJS(): Pojo {
    return {
      ...super.toJS(),
      connectors: this.connectors,
      isUpsideDown: this.isUpsideDown,
      parameters: this.parameters,
    };
  }
}

export default PathCard;
