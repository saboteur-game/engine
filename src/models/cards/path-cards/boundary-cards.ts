import { PathCard } from ".";
import { Sides } from "../card";
import getCardVisualization from "../../../utils/get-card-visualization";

export class StartPathCard extends PathCard {
  constructor(isUpsideDown?: boolean) {
    super([Sides.top, Sides.right, Sides.bottom, Sides.left], isUpsideDown);
  }

  visualize(): string {
    return getCardVisualization(this, "[]");
  }
}

export class FinishPathCard extends PathCard {}

export class GoldFinishPathCard extends FinishPathCard {
  constructor(isUpsideDown?: boolean) {
    super([Sides.top, Sides.right, Sides.bottom, Sides.left], isUpsideDown);
  }

  visualize(): string {
    return getCardVisualization(this, "!!");
  }
}

export class RockFinishPathCard extends FinishPathCard {
  visualize(): string {
    return getCardVisualization(this, "xx");
  }
}
