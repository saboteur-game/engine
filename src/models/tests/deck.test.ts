import { mocked } from "ts-jest/utils";
import Deck from "../deck";
import { getShuffledDeck } from "../cards";
import { PassageCard } from "../cards/path-cards";
import { Sides } from "../cards/card";

const mockedGetSuffledDeck = mocked(getShuffledDeck);

jest.mock("../cards");

describe("Deck", () => {
  let deck: Deck;
  let card: PassageCard;

  beforeEach(() => {
    card = new PassageCard([Sides.top, Sides.right, Sides.bottom, Sides.left]);
    mockedGetSuffledDeck.mockReturnValue([card]);
  });

  it("can be serialized", () => {
    expect(new Deck().toJS()).toMatchSnapshot();
  });

  describe("draw card when the pile is populated", () => {
    beforeEach(() => {
      deck = new Deck();
    });

    it("draws a card from the deck and returns it", () => {
      expect(deck.getCardCount()).toBe(1);
      expect(deck.drawCard()).toBe(card);
      expect(deck.getCardCount()).toBe(0);
    });
  });

  describe("draw card when the pile is empty", () => {
    beforeEach(() => {
      mockedGetSuffledDeck.mockReturnValue([]);
      deck = new Deck();
    });

    it("returns undefined", () => {
      expect(deck.getCardCount()).toBe(0);
      expect(deck.drawCard()).toBe(undefined);
      expect(deck.getCardCount()).toBe(0);
    });
  });
});
