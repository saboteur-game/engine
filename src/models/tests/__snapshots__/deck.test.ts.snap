// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deck can be serialized 1`] = `
Object {
  "drawPile": Array [
    PassageCard {
      "connectors": Array [
        1,
        2,
        3,
        4,
      ],
      "id": "test-id-0",
      "isUpsideDown": false,
      "status": "unused",
    },
  ],
  "id": "test-id-1",
}
`;
